internal/mode/static/state/graph/policies.go

```diff
@@ -0,0 +1,310 @@
+	Source policies.Policy
+	// Ancestor is the ancestor object of the Policy. Used in status.
+	Ancestor *PolicyAncestor
+	// TargetRef is the resource that the Policy targets.
+	TargetRef PolicyTargetRef
+	// Conditions holds the conditions for the Policy.
+	// These conditions apply to the entire Policy.
+	// The conditions in the Ancestor apply only to the Policy in regard to the Ancestor.
+	Conditions []conditions.Condition
+	// Valid indicates whether the Policy is valid.
+	Valid bool
+}
+
+// PolicyAncestor represents an ancestor of a Policy.
+type PolicyAncestor struct {
+	// Ancestor is the ancestor object.
+	Ancestor v1alpha2.ParentReference
+	// Conditions contains the list of conditions of the Policy in relation to the ancestor.
+	Conditions []conditions.Condition
+}
+
+// PolicyAncestorRef contains the identifying information of the ancestor.
+type PolicyAncestorRef struct {
+	// Nsname is the NamespacedName of the Ancestor.
+	Nsname types.NamespacedName
+	// Kind is the Kind of the Ancestor.
+	Kind v1.Kind
+	// Group is the Group of the Ancestor.
+	Group v1.Group
+}
+
+// PolicyTargetRef represents the object that the Policy is targeting.
+type PolicyTargetRef struct {
+	// Kind is the Kind of the object.
+	Kind v1.Kind
+	// Group is the Group of the object.
+	Group v1.Group
+	// Nsname is the NamespacedName of the object.
+	Nsname types.NamespacedName
+}
+
+// PolicyKey is a unique identifier for an NGF Policy.
+type PolicyKey struct {
+	// Nsname is the NamespacedName of the Policy.
+	NsName types.NamespacedName
+	// GVK is the GroupVersionKind of the Policy.
+	GVK schema.GroupVersionKind
+}
+
+const (
+	gatewayGroupKind = v1.GroupName + "/" + kinds.Gateway
+	hrGroupKind      = v1.GroupName + "/" + kinds.HTTPRoute
+	grpcGroupKind    = v1.GroupName + "/" + kinds.GRPCRoute
+)
+
+// attachPolicies attaches the graph's processed policies to the resources they target. It modifies the graph in place.
+func (g *Graph) attachPolicies(ctlrName string) {
+	if g.Gateway == nil {
+		return
+	}
+
+	for _, policy := range g.NGFPolicies {
+		ref := policy.TargetRef
+
+		switch ref.Kind {
+		case kinds.Gateway:
+			attachPolicyToGateway(policy, g.Gateway, g.IgnoredGateways, ctlrName)
+		case kinds.HTTPRoute, kinds.GRPCRoute:
+			route, exists := g.Routes[routeKeyForKind(ref.Kind, ref.Nsname)]
+			if !exists {
+				continue
+			}
+
+			attachPolicyToRoute(policy, route, ctlrName)
+		}
+	}
+}
+
+func attachPolicyToRoute(policy *Policy, route *L7Route, ctlrName string) {
+	kind := v1.Kind(kinds.HTTPRoute)
+	if route.RouteType == RouteTypeGRPC {
+		kind = kinds.GRPCRoute
+	}
+
+	routeNsName := types.NamespacedName{Namespace: route.Source.GetNamespace(), Name: route.Source.GetName()}
+
+	ancestor := &PolicyAncestor{
+		Ancestor: createParentReference(v1.GroupName, kind, routeNsName),
+	}
+
+	curAncestorStatus := policy.Source.GetPolicyStatus().Ancestors
+	if ancestorsFull(curAncestorStatus, ancestor.Ancestor, ctlrName) {
+		return
```

Comment:
I need to add a FIXME here. According to the spec, we need to communicate to the user that we can't implement the policy because the ancestor slice is full. Presumably, this means adding a condition to the ancestor's status.  I'm hesitant to do this because we could easily overflow the ancestor's conditions. I want to think about this a little more and come up with a design.

Note: this also applies to BackendTLSPolicies.

Reply:
There's also the case where different policy kinds (for example BackendTLSPolicy and ClientSettingsPolicy) have the same ancestor and a full ancestor status slice. That could also overflow the conditions of the ancestor if we were to add one condition per policy.

Reply:
Of course, that totally makes sense - I wasn't thinking at all 🤦🏼‍♀️ Thanks for the explanation!!

Reply:
Added a FIXME and an issue https://github.com/nginxinc/nginx-gateway-fabric/issues/1987